### PR TITLE
Fallback to GetFullPathNameW for Windows _canonicalizedPath

### PR DESCRIPTION
**Pull request by gmittert**
If a file is not found, rather than failing, we should use
GetFullPathNameW which doesn't resolve symlinks, but will work on paths
that don't exist.

### CODE REVIEW

**Comment by gmittert** [comment]
@swift-ci please test

File: Foundation/FileManager+Win32.swift [hunk 1]
```diff
@@ -312,7 +312,20 @@ extension FileManager {
                               nil)
         }
         if hFile == INVALID_HANDLE_VALUE {
-            throw _NSErrorWithWindowsError(GetLastError(), reading: true)
+            let error = GetLastError()
+            if error == ERROR_FILE_NOT_FOUND || error == ERROR_PATH_NOT_FOUND {
+                return try path.withCString(encodedAs: UTF16.self) {
+                    let dwLength = GetFullPathNameW($0, 0, nil, nil)
+                    var szPath = Array<WCHAR>(repeating: 0, count: Int(dwLength + 1))
+                    let result = GetFullPathNameW($0, dwLength + 1, &szPath, nil)
```
Comment (compnerd) [inline Foundation/FileManager+Win32.swift]: I think I prefer `GetFullPathNameW($0, szPath.count, &szPath, nil)`

File: Foundation/FileManager+Win32.swift [hunk 2]
```diff
@@ -312,7 +312,20 @@ extension FileManager {
...
+                    guard result != 0 && result <= dwLength + 1 else {
```
Comment (compnerd) [inline Foundation/FileManager+Win32.swift]: I think that `result <= szPath.count` is nicer, in fact, at that point you could change `let result` into `dwLength` and get something like:

`guard dwLength > 0 && dwLength <= szPath.count`

which feels more obvious

File: Foundation/FileManager+Win32.swift [hunk 3]
```diff
@@ -312,7 +312,20 @@ extension FileManager {
...
+                        throw _NSErrorWithWindowsError(GetLastError(), reading: true)
+                    }
+                    return String(decodingCString: szPath, as: UTF16.self)
+                }
+            } else {
```
Comment (compnerd) [inline Foundation/FileManager+Win32.swift]: I think that the else { can be elided. You will always get a result or a failure from the error recovery path.

**Comment by gmittert** [comment]
Updated for feedback :) It does seem nicer to use szPath.count.

@swift-ci please test

**Comment by gmittert** [comment]
@swift-ci please test and merge

**Comment by gmittert** [comment]
@swift-ci please test and merge

**Pull request merged**